src/zenml/models/v2/core/trigger.py

```diff
@@ -0,0 +1,364 @@
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Collection of all models concerning triggers."""
+import copy
+from typing import TYPE_CHECKING, Any, Dict, Optional, TypeVar, Union
+from uuid import UUID
+
+from pydantic import BaseModel, Field
+
+from zenml.constants import STR_FIELD_MAX_LENGTH
+from zenml.enums import PluginSubType
+from zenml.models.v2.base.base import BaseZenModel
+from zenml.models.v2.base.scoped import (
+    WorkspaceScopedFilter,
+    WorkspaceScopedRequest,
+    WorkspaceScopedResponse,
+    WorkspaceScopedResponseBody,
+    WorkspaceScopedResponseMetadata,
+    WorkspaceScopedResponseResources,
+)
+from zenml.models.v2.core.user import UserResponse
+
+if TYPE_CHECKING:
+    from zenml.models.v2.core.event_source import EventSourceResponse
+    from zenml.zen_stores.schemas import BaseSchema
+
+    AnySchema = TypeVar("AnySchema", bound=BaseSchema)
+
+# ------------------ Base Model ------------------
+
+
+class TriggerBase(BaseModel):
+    """Base model for triggers."""
+
+    name: str = Field(
+        title="The name of the Trigger.", max_length=STR_FIELD_MAX_LENGTH
+    )
+    description: str = Field(
+        default="",
+        title="The description of the trigger",
+        max_length=STR_FIELD_MAX_LENGTH,
+    )
+    event_source_id: UUID = Field(
+        title="The event source that activates this trigger.",
+    )
+    event_filter: Dict[str, Any] = Field(
+        title="Filter applied to events that activate this trigger.",
+    )
+
+    action: Dict[str, Any] = Field(
+        title="The configuration for the action that is executed by this "
+        "trigger.",
+    )
+    action_flavor: str = Field(
+        title="The flavor of the action that is executed by this trigger.",
+        max_length=STR_FIELD_MAX_LENGTH,
+    )
+    action_subtype: PluginSubType = Field(
+        title="The subtype of the action that is executed by this trigger.",
+        max_length=STR_FIELD_MAX_LENGTH,
+    )
+    service_account_id: UUID = Field(
+        title="The service account that is used to execute the action.",
+    )
+    auth_window: Optional[int] = Field(
+        default=None,
+        title="The time window in minutes for which the service account is "
+        "authorized to execute the action. Set this to 0 to authorize the "
+        "service account indefinitely (not recommended). If not set, a "
+        "default value defined for each individual action type is used.",
+    )
+
+
+# ------------------ Request Model ------------------
+class TriggerRequest(TriggerBase, WorkspaceScopedRequest):
+    """Model for creating a new Trigger."""
+
+    # executions: somehow we need to link to executed Actions here
```

Comment:
This is not a TODO, just adding a comment here so it won't be missed

Reply:
@AlexejPenner should we return the last X (10?) executions as trigger.resources ?

Reply:
yes I think so, but not in the Request model.

Reply:
scratch that, I went ahead and actually implemented it, wasn't too difficult